search/src/kernels.cu

```diff
@@ -30,26 +29,27 @@ __global__ void convolvePSF(int width, int height,
     const int x = blockIdx.x*CONV_THREAD_DIM+threadIdx.x;
     const int y = blockIdx.y*CONV_THREAD_DIM+threadIdx.y;
     if (x < 0 || x > width-1 || y < 0 || y > height-1) return;
-    const int minX = max(x-psfRad, 0);
-    const int minY = max(y-psfRad, 0);
-    const int maxX = min(x+psfRad, width-1);
-    const int maxY = min(y+psfRad, height-1);
 
     // Read kernel
     float sum = 0.0;
     float psfPortion = 0.0;
     float center = sourceImage[y*width+x];
     if (center != NO_DATA) {
-        for (int j=minY; j<=maxY; j++)
+        for (int j = -psfRad; j <= psfRad; j++)
         {
             // #pragma unroll
-            for (int i=minX; i<=maxX; i++)
+            for (int i = -psfRad; i <= psfRad; i++)
             {
-                float currentPixel = sourceImage[j*width+i];
-                if (currentPixel != NO_DATA) {
-                    float currentPSF = psf[(j-minY)*psfDim+i-minX];
-                    psfPortion += currentPSF;
-                    sum += currentPixel * currentPSF;
+                if ((x + i >= 0) && (x + i < width) &&
```

Comment:
In general good to avoid branching in GPU kernels when possible but in the common case all threads will be taking the same path anyway so it shouldn't really matter. And convolution won't be a performance bottleneck for the whole pipeline anyway as long as it's reasonably performant.

Reply:
Thanks for the heads up about branching in GPUs. I will try to think of alternate ways to do this type of thing in the future.